be/src/io/cache/file_cache_manager.cpp

```diff
@@ -56,88 +80,82 @@ void FileCacheManager::remove_file_cache(const std::string& cache_path) {
+                // check if it is a dir name
+                if (ends_with(seg_filename, ".dat")) {
                     continue;
                 }
-                if (!io::global_local_filesystem()->delete_file(cache_file_path).ok()) {
-                    LOG(ERROR) << "delete_file failed: " << cache_file_path;
+                // skip file cache already in memory
+                std::stringstream ss;
+                ss << tablet->tablet_path() << "/" << seg_filename;
+                std::string cache_path = ss.str();
+                if (_file_cache_map.find(cache_path) != _file_cache_map.end()) {
                     continue;
                 }
+
+                auto file_cache = std::make_shared<DummyFileCache>(
+                        cache_path, config::file_cache_alive_time_sec);
+                // load cache meta from disk and clean unfinished cache files
+                file_cache->load_and_clean();
+                // policy1: GC file cache by timeout
+                file_cache->clean_timeout_cache();
+                // sort file cache by last match time
+                _add_file_cache_for_gc_by_disk(contexts, file_cache);
             }
-            // find cache file without done file.
-            for (std::list<std::string>::iterator itr = done_names.begin(); itr != done_names.end();
-                 ++itr) {
-                std::string cache_filename = StringReplace(*itr, CACHE_DONE_FILE_SUFFIX, "", true);
-                if (cache_names.find(cache_filename) != cache_names.end()) {
-                    cache_names.erase(cache_filename);
-                }
-            }
-            // remove cache file without done file
-            for (std::map<std::string, bool>::iterator itr = cache_names.begin();
-                 itr != cache_names.end(); ++itr) {
-                std::stringstream cache_file_ss;
-                cache_file_ss << cache_path << "/" << itr->first;
-                std::string cache_file_path = cache_file_ss.str();
-                time_t m_time;
-                if (!FileUtils::mtime(cache_file_path, &m_time).ok()) {
-                    continue;
-                }
-                if (now - m_time < config::file_cache_alive_time_sec) {
-                    continue;
-                }
-                LOG(INFO) << "Delete cache file without done file: " << cache_file_path;
-                if (!io::global_local_filesystem()->delete_file(cache_file_path).ok()) {
-                    LOG(ERROR) << "delete_file failed: " << cache_file_path;
-                }
-            }
-            if (io::global_local_filesystem()->list(cache_path, &cache_file_names).ok() &&
-                cache_file_names.size() == 0) {
-                if (global_local_filesystem()->delete_directory(cache_path).ok()) {
-                    LOG(INFO) << "Delete empty dir: " << cache_path;
-                }
-            }
+        }
+    }
+
+    // process file caches in memory
```

Comment:
You need add a read lock for _file_cache_map by call std::shared_lock<std::shared_mutex> rdlock(_cache_map_lock);

Reply:
line 114 already added rdlock, and the scope includes these lines.